8360886: Cmd + plus shortcut does not work reliably.

beldenfox opened the pull request:
The Mac platform code figures out where characters are on the keyboard as the user types. The character table is updated on every key press by calling a registerKeyEvent: in GlassApplication. This character table is used to resolve KeyCharacterCombination accelerators like Cmd + "+".

On a US English layout when the user types Cmd + "+" on the main keyboard they're actually typing Cmd + "=". There's special handling in macOS for this combination that can cause two NSEvents to be sent and so there's special handling in the Glass code so we don't process both events. When this special case is invoked registerKeyEvent: isn't being called to update the character table. This bug was introduced when code was consolidated in PR #1528.

The fix is a simple one-liner. I've added a test for this in the KeyboardTest.java program. It's an isolated test because it requires the Robot to send events to hold down a modifier while a character key is pressed. I also updated some obsolete comments and tweaked it to test all KeyCharacterCombinations since they should now work reliably on all platforms.

andy-goryachev-oracle commented:
/reviewers 2

andy-goryachev-oracle commented on tests/manual/events/KeyboardTest.java:
```diff
@@ -808,6 +807,105 @@ private void cleanup() {
+    // Cmd + "+" works as expected on the US English layout on the Mac. The
+    // OS special-cases this combination and so does JavaFX. In any case
+    // Shortcut + "+" is a very common shortcut and deserves a test.
+    private void testShiftedShortcut(Layout layout, Node focusNode, Logging log) {
```
Question: is it possible to convert this manual test into an automated headful one?

andy-goryachev-oracle replied:
I was thinking more in terms of exhaustive checking of all the available keys (so we won't encounter another scenario when a bug is introduced similar to #1528, if that is possible).

I know we can't really control the keyboard layout, so perhaps what I am asking may not be possible since these tests might fail on other layouts.

So let me ask
- is the problem limited to cmd+= on the US keyboard?
- could we have a different behavior with a different kb layout (German? French?)
- what happens when the user attaches an external USB or bluetooth keyboard?

beldenfox replied:
The issue here is limited to cmd+=  and cmd+period. These receive special treatment by macOS. To counter-act this they also receive special treatment in the Glass code.

The macOS special handling of cmd+= only applies to certain keyboard layouts (for example, English but not German). But that's not true for Glass, the work-around in glass applies to all layouts.

I will look into writing up a system test. I think I can put one together that tests more keys and works on all layouts. It would be nice not to rely solely on this manual test.

External USB or bluetooth keyboards shouldn't make a difference.

(In case you're curious, Apple wants to make sure you can invoke Cmd + "+" by holding down Cmd and pressing any key that has a "+" symbol printed on it. On a US English layout that means Cmd + "=" should also work. So they send us Cmd + "=" and if that doesn't trigger a shortcut they send us Cmd + "+". Glass tries to ensure that that second event doesn't turn into another KeyEvent. I think this only happens with keys where "+" is Shift + "=" which is why it happens on US English but not German.)

andy-goryachev-oracle replied:
thank you for the detailed explanations!

andy-goryachev-oracle commented:
SKARA warns about trailing period in the title.  I've corrected the JIRA ticket, please fix the PR.

andy-goryachev-oracle commented on tests/manual/events/KeyboardTest.java:
```diff
@@ -38,6 +39,7 @@
 import javafx.scene.control.TextArea;
 import javafx.scene.input.KeyCharacterCombination;
 import javafx.scene.input.KeyCode;
+import javafx.scene.input.KeyCombination;
```
I wonder if it is possible to add a panel to the test stage with explanations that are currently in the class-level javadoc?  Specifically, the fact that you need to configure a particular layout.

I had some work done to simplify the boilerplate in #1747, though it did not make it yet.

beldenfox commented:
> SKARA warns about trailing period in the title. I've corrected the JIRA ticket, please fix the PR.

Title updated. Thanks for fixing the ticket.

beldenfox commented:
Well, now I’ve gone and re-written this PR.

As stated earlier sometimes a single key press can lead to multiple calls to performKeyEquivalent. During testing I discovered that the way Glass checks for this doesn’t always work. Currently it tries to guess which key event might lead to a second one but there are cases where it guesses wrong (e.g. the Hebrew layout). I’ve decided to take a different approach, detecting and discarding the second event. The logic is:

- Is the new key event not the current event (theEvent != NSApp.currentEvent)?
- Are we still working on an earlier event (NSApp.currentEvent == lastKeyEvent)?

If both these conditions are true we can ignore the new key event.

I’ve reverted the changes to the manual test and added a system test. It covers a smattering of carefully selected keys including the combinations involved in this bug. Currently it’s restricted to macOS and Latin keyboards because the Hebrew keyboard on macOS is crazy. I will expand this to other platforms when I have time to do the testing. I’ve run this test on macOS using US English, French, German, Spanish, Hebrew, Devanagari, and a few other layouts.

(About that Hebrew layout. When you hold down Command all the keys start behaving like a US English layout, including the punctuation keys. So the key you press to type a period is entirely different from the key that generates Cmd+period.)

mstr2 reviewed (approved):
Looks good.

andy-goryachev-oracle reviewed (approved):
Sorry, I've dropped the ball with the re-review.  lgtm.

kevinrushforth commented:
@beldenfox In case you missed it, this is ready for integration.

beldenfox commented:
/integrate
